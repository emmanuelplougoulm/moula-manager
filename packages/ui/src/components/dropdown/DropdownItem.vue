<template>
  <div class="container">
    <LabelItem
      class="label"
      :data-testid="`label`"
      :text="label"
      :highlighted="!selected"
      :tooltip-title="props.tooltipTitle"
      :tooltip-description="props.tooltipDescription"
    />
    <div class="select-container">
      <div
        :data-testid="`select`"
        class="text-big-regular select"
        :class="selectClasses"
        @click="toggle"
      >
        <span class="select-text">
          {{ selectedText }}
        </span>
        <KeyboardArrowDown class="select-icon" :class="iconClass" />
      </div>
      <div v-show="opened" class="options">
        <div
          v-for="option in options"
          :key="option.id"
          class="text-big-regular option"
          :data-testid="`option-${option.id}`"
          @click="selectOption(option)"
        >
          <Check v-show="showOptionCheck(option.id)" class="option-icon" />
          <span class="option-text">{{ option.text }}</span>
        </div>
      </div>
    </div>
    <ErrorItem :text="errorText" />
  </div>
</template>

<script setup lang="ts">
import { ref, computed, defineExpose } from 'vue';
import KeyboardArrowDown from '../icons/KeyboardArrowDown.vue';
import Check from '../icons/Check.vue';
import ErrorItem from '../error/ErrorItem.vue';
import LabelItem from '../label/LabelItem.vue';
import type { DropdownOptionType } from '../../types';
// import useDetectOutsideClick from '../../composables/useDetectOutsideClick';

const props = defineProps({
  placeholder: {
    type: String,
    required: true
  },
  modelValue: {
    type: Object,
    required: true
  },
  label: {
    type: String,
    required: true
  },
  options: {
    type: Array as () => DropdownOptionType[],
    required: true
  },
  error: {
    type: String,
    default: undefined
  },
  tooltipTitle: {
    type: String,
    default: undefined
  },
  tooltipDescription: {
    type: String,
    default: undefined
  }
});

const selectedOption = ref(props.modelValue);
const selectedText = computed(() => selectedOption.value.text ?? props.placeholder);
const selected = computed(() => Object.keys(selectedOption.value).length > 0);
const opened = ref(false);
const errorText = ref(props.error);

const selectClasses = computed(() => ({
  ['selected-colors selected-border']: selected.value
}));
const iconClass = computed(() => ({
  ['selected-colors']: selected.value
}));

const showOptionCheck = (id: number) => {
  return id === selectedOption.value.id;
};
const toggle = () => {
  opened.value = !opened.value;
  console.log('am inside');
};

function closeModal() {
  // opened.value = false;
  console.log('am triggered');
}

defineExpose({ closeModal });

// const select = ref(null);
// console.log('selectRef');
// useDetectOutsideClick(select, closemodal);

const emit = defineEmits(['update:modelValue']);
const selectOption = (option: DropdownOptionType) => {
  errorText.value = '';
  selectedOption.value = option;
  opened.value = false;
  emit('update:modelValue', option);
};
</script>

<style scoped>
.label {
  z-index: 0;
}
.container {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}
.select-container {
  max-height: 3.25rem;
}

:is(.select, .option) {
  box-sizing: border-box;
  cursor: pointer;
}

.select {
  display: flex;
  justify-content: space-between;
  align-items: center;
  border-radius: 0.25rem;
  padding: 0.75rem 1rem;
  border: 0.125rem solid var(--primary-30);
  background-color: var(--primary-90);
  color: var(--primary-text);
  position: relative;
  z-index: 0;
}

.select-text {
  padding-right: 1rem;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}

.select-icon {
  color: var(--primary-30);
  width: 1.5rem;
  height: 1.5rem;
}

.selected-colors {
  background-color: var(--neutral-80);
  color: var(--neutral-20);
}

.selected-border {
  border: 0.063rem solid var(--neutral-60);
}

.options {
  z-index: 99;
  max-height: 21rem;
  overflow-y: scroll;
  color: var(--neutral-20);
  border-radius: 0rem 0rem 0.25rem 0.25rem;
  box-shadow: var(--box-shadow);
  background-color: var(--neutral-100);
  position: relative;
}

.option {
  padding: 0.75rem 1rem 0.75rem 0.5rem;
  display: grid;
  grid-template-columns: 1.25rem 1fr;
  grid-template-areas: 'icon text';
  align-items: center;
  gap: 0.5rem;
  line-height: 1.375rem;
}
.option:hover {
  background-color: var(--neutral-80);
}

.option:not(:first-child) {
  border-top: 0.063rem solid var(--neutral-75);
}

.option-icon {
  grid-area: icon;
  color: var(--neutral-20);
  width: 1.25rem;
  height: 1.25rem;
}

.option-text {
  grid-area: text;
}
</style>
