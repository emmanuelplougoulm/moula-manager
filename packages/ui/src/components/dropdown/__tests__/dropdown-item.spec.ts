import { mount } from '@vue/test-utils';
import DropdownItem from '../DropdownItem.vue';

const placeholder = 'SÃ©lectionnez une option';
const requiredText = 'This field is required';

const options = [
  {
    id: 1,
    text: 'Option 1: Lorem ipsum dolor sit amet.',
    value: 'option-1'
  },
  {
    id: 2,
    text: 'Option 2: Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam erat odio, accumsan non risus sed, imperdiet malesuada diam. Etiam dictum mauris sed malesuada volutpat. Morbi accumsan vitae enim ut facilisis.',
    value: 'option-2'
  }
];

describe('DropdownItem component tests', () => {
  it('should display the component with a placeholder when no option is selected', () => {
    const wrapper = mount(DropdownItem, {
      props: {
        label: 'Text',
        options: options,
        placeholder: placeholder,
        modelValue: {}
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display the component with a selected option and emit a update:modelValue event', async () => {
    const wrapper = mount(DropdownItem, {
      props: {
        label: 'Text',
        options: options,
        placeholder: placeholder,
        modelValue: {}
      }
    });
    const option = wrapper.find("[data-testid='option-1']");
    await option.trigger('click');

    expect(wrapper.emitted()).toHaveProperty('update:modelValue');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display an error message if the error prop is not empty', async () => {
    const wrapper = mount(DropdownItem, {
      props: {
        label: 'Text',
        error: requiredText,
        options: options,
        placeholder: placeholder,
        modelValue: {}
      }
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should not display an error message if an option is selected', async () => {
    const wrapper = mount(DropdownItem, {
      props: {
        label: 'Text',
        error: requiredText,
        options: options,
        placeholder: placeholder,
        modelValue: {}
      }
    });
    const option = wrapper.find("[data-testid='option-1']");
    await option.trigger('click');

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should close dropdown on click outside', async () => {
    const wrapper = mount(DropdownItem, {
      props: {
        label: 'Text',
        error: requiredText,
        options: options,
        placeholder: placeholder,
        modelValue: {}
      }
    });

    const select = wrapper.find("[data-testid='select']");
    await select.trigger('click');

    const label = wrapper.find("[data-testid='label']");
    await label.trigger('click');

    expect(wrapper.html()).toMatchSnapshot();
  });
});
