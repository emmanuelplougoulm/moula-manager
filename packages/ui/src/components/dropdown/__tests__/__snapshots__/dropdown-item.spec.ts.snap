// Vitest Snapshot v1

exports[`DropdownItem component tests > should close dropdown on click outside 1`] = `
"<div data-v-124a8955=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-124a8955=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted label\\" data-testid=\\"label\\">Text <!--v-if--></span>
  <div data-v-124a8955=\\"\\" class=\\"select-container\\">
    <div data-v-124a8955=\\"\\" data-testid=\\"select\\" class=\\"text-big-regular select\\"><span data-v-124a8955=\\"\\" class=\\"select-text\\">Sélectionnez une option</span>
      <div data-v-b9dd0015=\\"\\" data-v-124a8955=\\"\\" class=\\"container select-icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M480 725.566 226.434 472 283 415.434l197 197 197-197L733.566 472 480 725.566Z\\"></path>
        </svg></div>
    </div>
    <div data-v-124a8955=\\"\\" class=\\"options\\" style=\\"\\">
      <div data-v-124a8955=\\"\\" class=\\"text-big-regular option\\" data-testid=\\"option-1\\">
        <div data-v-b9dd0015=\\"\\" data-v-124a8955=\\"\\" class=\\"container option-icon\\" style=\\"display: none;\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div><span data-v-124a8955=\\"\\" class=\\"option-text\\">Option 1: Lorem ipsum dolor sit amet.</span>
      </div>
      <div data-v-124a8955=\\"\\" class=\\"text-big-regular option\\" data-testid=\\"option-2\\">
        <div data-v-b9dd0015=\\"\\" data-v-124a8955=\\"\\" class=\\"container option-icon\\" style=\\"display: none;\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div><span data-v-124a8955=\\"\\" class=\\"option-text\\">Option 2: Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam erat odio, accumsan non risus sed, imperdiet malesuada diam. Etiam dictum mauris sed malesuada volutpat. Morbi accumsan vitae enim ut facilisis.</span>
      </div>
    </div>
  </div><span data-v-c1d8433f=\\"\\" data-v-124a8955=\\"\\" id=\\"error-item\\" data-testid=\\"error-item\\" class=\\"text-small-regular\\">This field is required</span>
</div>"
`;

exports[`DropdownItem component tests > should display an error message if the error prop is not empty 1`] = `
"<div data-v-124a8955=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-124a8955=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted label\\" data-testid=\\"label\\">Text <!--v-if--></span>
  <div data-v-124a8955=\\"\\" class=\\"select-container\\">
    <div data-v-124a8955=\\"\\" data-testid=\\"select\\" class=\\"text-big-regular select\\"><span data-v-124a8955=\\"\\" class=\\"select-text\\">Sélectionnez une option</span>
      <div data-v-b9dd0015=\\"\\" data-v-124a8955=\\"\\" class=\\"container select-icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M480 725.566 226.434 472 283 415.434l197 197 197-197L733.566 472 480 725.566Z\\"></path>
        </svg></div>
    </div>
    <div data-v-124a8955=\\"\\" class=\\"options\\" style=\\"display: none;\\">
      <div data-v-124a8955=\\"\\" class=\\"text-big-regular option\\" data-testid=\\"option-1\\">
        <div data-v-b9dd0015=\\"\\" data-v-124a8955=\\"\\" class=\\"container option-icon\\" style=\\"display: none;\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div><span data-v-124a8955=\\"\\" class=\\"option-text\\">Option 1: Lorem ipsum dolor sit amet.</span>
      </div>
      <div data-v-124a8955=\\"\\" class=\\"text-big-regular option\\" data-testid=\\"option-2\\">
        <div data-v-b9dd0015=\\"\\" data-v-124a8955=\\"\\" class=\\"container option-icon\\" style=\\"display: none;\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div><span data-v-124a8955=\\"\\" class=\\"option-text\\">Option 2: Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam erat odio, accumsan non risus sed, imperdiet malesuada diam. Etiam dictum mauris sed malesuada volutpat. Morbi accumsan vitae enim ut facilisis.</span>
      </div>
    </div>
  </div><span data-v-c1d8433f=\\"\\" data-v-124a8955=\\"\\" id=\\"error-item\\" data-testid=\\"error-item\\" class=\\"text-small-regular\\">This field is required</span>
</div>"
`;

exports[`DropdownItem component tests > should display the component with a placeholder when no option is selected 1`] = `
"<div data-v-124a8955=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-124a8955=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted label\\" data-testid=\\"label\\">Text <!--v-if--></span>
  <div data-v-124a8955=\\"\\" class=\\"select-container\\">
    <div data-v-124a8955=\\"\\" data-testid=\\"select\\" class=\\"text-big-regular select\\"><span data-v-124a8955=\\"\\" class=\\"select-text\\">Sélectionnez une option</span>
      <div data-v-b9dd0015=\\"\\" data-v-124a8955=\\"\\" class=\\"container select-icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M480 725.566 226.434 472 283 415.434l197 197 197-197L733.566 472 480 725.566Z\\"></path>
        </svg></div>
    </div>
    <div data-v-124a8955=\\"\\" class=\\"options\\" style=\\"display: none;\\">
      <div data-v-124a8955=\\"\\" class=\\"text-big-regular option\\" data-testid=\\"option-1\\">
        <div data-v-b9dd0015=\\"\\" data-v-124a8955=\\"\\" class=\\"container option-icon\\" style=\\"display: none;\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div><span data-v-124a8955=\\"\\" class=\\"option-text\\">Option 1: Lorem ipsum dolor sit amet.</span>
      </div>
      <div data-v-124a8955=\\"\\" class=\\"text-big-regular option\\" data-testid=\\"option-2\\">
        <div data-v-b9dd0015=\\"\\" data-v-124a8955=\\"\\" class=\\"container option-icon\\" style=\\"display: none;\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div><span data-v-124a8955=\\"\\" class=\\"option-text\\">Option 2: Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam erat odio, accumsan non risus sed, imperdiet malesuada diam. Etiam dictum mauris sed malesuada volutpat. Morbi accumsan vitae enim ut facilisis.</span>
      </div>
    </div>
  </div>
  <!--v-if-->
</div>"
`;

exports[`DropdownItem component tests > should display the component with a selected option and emit a update:modelValue event 1`] = `
"<div data-v-124a8955=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-124a8955=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label\\" data-testid=\\"label\\">Text <!--v-if--></span>
  <div data-v-124a8955=\\"\\" class=\\"select-container\\">
    <div data-v-124a8955=\\"\\" data-testid=\\"select\\" class=\\"text-big-regular select selected-colors selected-border\\"><span data-v-124a8955=\\"\\" class=\\"select-text\\">Option 1: Lorem ipsum dolor sit amet.</span>
      <div data-v-b9dd0015=\\"\\" data-v-124a8955=\\"\\" class=\\"container select-icon selected-colors\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M480 725.566 226.434 472 283 415.434l197 197 197-197L733.566 472 480 725.566Z\\"></path>
        </svg></div>
    </div>
    <div data-v-124a8955=\\"\\" class=\\"options\\" style=\\"display: none;\\">
      <div data-v-124a8955=\\"\\" class=\\"text-big-regular option\\" data-testid=\\"option-1\\">
        <div data-v-b9dd0015=\\"\\" data-v-124a8955=\\"\\" class=\\"container option-icon\\" style=\\"\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div><span data-v-124a8955=\\"\\" class=\\"option-text\\">Option 1: Lorem ipsum dolor sit amet.</span>
      </div>
      <div data-v-124a8955=\\"\\" class=\\"text-big-regular option\\" data-testid=\\"option-2\\">
        <div data-v-b9dd0015=\\"\\" data-v-124a8955=\\"\\" class=\\"container option-icon\\" style=\\"display: none;\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div><span data-v-124a8955=\\"\\" class=\\"option-text\\">Option 2: Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam erat odio, accumsan non risus sed, imperdiet malesuada diam. Etiam dictum mauris sed malesuada volutpat. Morbi accumsan vitae enim ut facilisis.</span>
      </div>
    </div>
  </div>
  <!--v-if-->
</div>"
`;

exports[`DropdownItem component tests > should not display an error message if an option is selected 1`] = `
"<div data-v-124a8955=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-124a8955=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label\\" data-testid=\\"label\\">Text <!--v-if--></span>
  <div data-v-124a8955=\\"\\" class=\\"select-container\\">
    <div data-v-124a8955=\\"\\" data-testid=\\"select\\" class=\\"text-big-regular select selected-colors selected-border\\"><span data-v-124a8955=\\"\\" class=\\"select-text\\">Option 1: Lorem ipsum dolor sit amet.</span>
      <div data-v-b9dd0015=\\"\\" data-v-124a8955=\\"\\" class=\\"container select-icon selected-colors\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M480 725.566 226.434 472 283 415.434l197 197 197-197L733.566 472 480 725.566Z\\"></path>
        </svg></div>
    </div>
    <div data-v-124a8955=\\"\\" class=\\"options\\" style=\\"display: none;\\">
      <div data-v-124a8955=\\"\\" class=\\"text-big-regular option\\" data-testid=\\"option-1\\">
        <div data-v-b9dd0015=\\"\\" data-v-124a8955=\\"\\" class=\\"container option-icon\\" style=\\"\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div><span data-v-124a8955=\\"\\" class=\\"option-text\\">Option 1: Lorem ipsum dolor sit amet.</span>
      </div>
      <div data-v-124a8955=\\"\\" class=\\"text-big-regular option\\" data-testid=\\"option-2\\">
        <div data-v-b9dd0015=\\"\\" data-v-124a8955=\\"\\" class=\\"container option-icon\\" style=\\"display: none;\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div><span data-v-124a8955=\\"\\" class=\\"option-text\\">Option 2: Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam erat odio, accumsan non risus sed, imperdiet malesuada diam. Etiam dictum mauris sed malesuada volutpat. Morbi accumsan vitae enim ut facilisis.</span>
      </div>
    </div>
  </div>
  <!--v-if-->
</div>"
`;
